File: leeaa/173.py
class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """ 
        self.stack = []
        if not root:
            return

        self.stack=[root]
        while self.stack[-1].left:
            self.stack.append(self.stack[-1].left)

    def hasNext(self):
        """
        :rtype: bool
        """
        if len(self.stack):
            return True
        else:
            return False

    def next(self):
        """
        :rtype: int
        """
        ret_node = self.stack.pop()
        if ret_node.right:
            self.stack.append(ret_node.right)
            while self.stack[-1].left:
                self.stack.append(self.stack[-1].left)
        return ret_node.val

File: rein/173.js
/**
 * Definition for binary tree
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @constructor
 * @param {TreeNode} root - root of the binary search tree
 */
var BSTIterator = function(root) {
    this.stack = [];
    this._pushLeftEdge(root);
};

BSTIterator.prototype._pushLeftEdge = function (node) {
    for (let t = node; t != null; t = t.left) {
        this.stack.push(t);
    }
}


/**
 * @this BSTIterator
 * @returns {boolean} - whether we have a next smallest number
 */
BSTIterator.prototype.hasNext = function() {
    return this.stack.length > 0;
};

/**
 * @this BSTIterator
 * @returns {number} - the next smallest number
 */
BSTIterator.prototype.next = function() {
    if (!this.hasNext()) {
        throw new Error('no more');
    }
    const node = this.stack.pop();
    this._pushLeftEdge(node.right);
    return node.val;
};

/**
 * Your BSTIterator will be called like this:
 * var i = new BSTIterator(root), a = [];
 * while (i.hasNext()) a.push(i.next());
*/
